File: apps/web/src/domain/catalogs/gameplayCatalog.ts
import gameplayCatalog from "../../../../../packages/ui-domain/gameplay_catalog.json";

type CharacterFaceCatalog = {
  face_id: string;
  name: string;
  aliases: string[];
};

type CharacterSlotCatalog = {
  slot: number;
  card_no: number;
  faces: [CharacterFaceCatalog, CharacterFaceCatalog];
};

type WeatherCatalogItem = {
  weather_id: number;
  name: string;
  effect: string;
};

type GameplayCatalog = {
  character_slots: CharacterSlotCatalog[];
  weather_cards: WeatherCatalogItem[];
};

const catalog = gameplayCatalog as GameplayCatalog;

const slotByAlias = new Map<string, number>();
const facesBySlot = new Map<number, [string, string]>();
const weatherByName = new Map<string, WeatherCatalogItem>();

for (const slot of catalog.character_slots) {
  facesBySlot.set(slot.slot, [slot.faces[0].name, slot.faces[1].name]);
  for (const face of slot.faces) {
    slotByAlias.set(face.name.trim(), slot.slot);
    for (const alias of face.aliases) {
      slotByAlias.set(alias.trim(), slot.slot);
    }
  }
}

for (const weather of catalog.weather_cards) {
  weatherByName.set(weather.name, weather);
}

export function prioritySlotForCharacterName(character: string | null | undefined): number | null {
  if (typeof character !== "string") {
    return null;
  }
  return slotByAlias.get(character.trim()) ?? null;
}

export function charactersForSlot(slot: number): [string, string] | null {
  return facesBySlot.get(slot) ?? null;
}

export function oppositeCharacterNameForSlot(slot: number, activeCharacter: string | null | undefined): string | null {
  const pair = charactersForSlot(slot);
  if (!pair) {
    return null;
  }
  if (typeof activeCharacter !== "string" || !activeCharacter.trim()) {
    return pair[1];
  }
  const normalized = activeCharacter.trim();
  if (pair[0] === normalized) {
    return pair[1];
  }
  if (pair[1] === normalized) {
    return pair[0];
  }
  return pair[1];
}

export function weatherByDisplayName(name: string | null | undefined): WeatherCatalogItem | null {
  if (typeof name !== "string" || !name.trim()) {
    return null;
  }
  return weatherByName.get(name.trim()) ?? null;
}

export function weatherEffectForDisplayName(name: string | null | undefined): string | null {
  return weatherByDisplayName(name)?.effect ?? null;
}

export { catalog as gameplayCatalog };

File: apps/server/src/domain/gameplay_catalog.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any


def _catalog_path() -> Path:
    return Path(__file__).resolve().parents[4] / "packages" / "ui-domain" / "gameplay_catalog.json"


@lru_cache(maxsize=1)
def load_gameplay_catalog() -> dict[str, Any]:
    return json.loads(_catalog_path().read_text(encoding="utf-8"))


@lru_cache(maxsize=1)
def _slot_by_alias() -> dict[str, int]:
    mapping: dict[str, int] = {}
    for slot in load_gameplay_catalog()["character_slots"]:
        slot_no = int(slot["slot"])
        for face in slot["faces"]:
            name = str(face["name"]).strip()
            if name:
                mapping[name] = slot_no
            for alias in face.get("aliases", []):
                normalized = str(alias).strip()
                if normalized:
                    mapping[normalized] = slot_no
    return mapping


@lru_cache(maxsize=1)
def _faces_by_slot() -> dict[int, tuple[str, str]]:
    mapping: dict[int, tuple[str, str]] = {}
    for slot in load_gameplay_catalog()["character_slots"]:
        faces = slot["faces"]
        mapping[int(slot["slot"])] = (str(faces[0]["name"]), str(faces[1]["name"]))
    return mapping


def priority_slot_for_character_name(character: str | None) -> int | None:
    if not isinstance(character, str):
        return None
    normalized = character.strip()
    if not normalized:
        return None
    return _slot_by_alias().get(normalized)


def characters_for_slot(slot: int) -> tuple[str, str] | None:
    return _faces_by_slot().get(slot)


def opposite_character_name_for_slot(slot: int, active_character: str | None) -> str | None:
    pair = characters_for_slot(slot)
    if pair is None:
        return None
    normalized = active_character.strip() if isinstance(active_character, str) else ""
    if not normalized:
        return pair[1]
    if pair[0] == normalized:
        return pair[1]
    if pair[1] == normalized:
        return pair[0]
    return pair[1]
